File: fetch-matches.py
#!/usr/bin/env python3
"""
Script to fetch Premier League match results from the official API
and output them as a JSON file.

Usage:
    python3 fetch-matches.py [--matchweek WEEK] [--all] [--output FILE]

Examples:
    # Fetch matchweek 15 (default)
    python3 fetch-matches.py
    
    # Fetch specific matchweek
    python3 fetch-matches.py --matchweek 10
    
    # Fetch all matchweeks (1-38)
    python3 fetch-matches.py --all
    
    # Save to file
    python3 fetch-matches.py --all --output matches.json
"""

import argparse
import json
import sys
import urllib.request
import urllib.error

def fetch_matches(matchweek):
    """
    Fetch matches for a specific matchweek from the Premier League API.
    
    Args:
        matchweek: The matchweek number to fetch
        
    Returns:
        dict: The API response as a dictionary, or None if error
    """
    url = f"https://sdp-prem-prod.premier-league-prod.pulselive.com/api/v2/matches?competition=8&season=2024&matchweek={matchweek}&_limit=100"
    
    # Create request with headers to mimic a browser
    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36',
        'Accept': 'application/json',
        'Accept-Language': 'en-US,en;q=0.9',
    }
    
    req = urllib.request.Request(url, headers=headers)
    
    try:
        with urllib.request.urlopen(req, timeout=10) as response:
            data = response.read()
            return json.loads(data)
    except urllib.error.URLError as e:
        print(f"Error fetching matchweek {matchweek}: {e}", file=sys.stderr)
        return None
    except json.JSONDecodeError as e:
        print(f"Error parsing JSON for matchweek {matchweek}: {e}", file=sys.stderr)
        return None

def main():
    """
    Main function to fetch matchweeks and output results.
    """
    parser = argparse.ArgumentParser(
        description='Fetch Premier League match results from the official API'
    )
    parser.add_argument(
        '--matchweek', 
        type=int, 
        default=15,
        help='Matchweek number to fetch (default: 15)'
    )
    parser.add_argument(
        '--all',
        action='store_true',
        help='Fetch all matchweeks (1-38)'
    )
    parser.add_argument(
        '--output',
        type=str,
        help='Output file path (default: stdout)'
    )
    
    args = parser.parse_args()
    
    if args.all:
        # Fetch all matchweeks
        print("Fetching all matchweeks (1-38)...", file=sys.stderr)
        all_data = []
        
        for week in range(1, 39):
            print(f"Fetching matchweek {week}...", file=sys.stderr)
            data = fetch_matches(week)
            
            if data is not None:
                all_data.append({
                    'matchweek': week,
                    'data': data
                })
            else:
                print(f"Warning: Failed to fetch matchweek {week}", file=sys.stderr)
        
        result = {
            'season': 2024,
            'competition': 8,
            'matchweeks': all_data
        }
    else:
        # Fetch single matchweek
        print(f"Fetching matchweek {args.matchweek}...", file=sys.stderr)
        result = fetch_matches(args.matchweek)
        
        if result is None:
            print("Failed to fetch data. The API might be unavailable, blocked, or require authentication.", file=sys.stderr)
            print("\nNote: If you're running this in a restricted environment, the domain might be blocked.", file=sys.stderr)
            print("Try running this script from a different network or machine.", file=sys.stderr)
            sys.exit(1)
    
    # Output the JSON data
    json_output = json.dumps(result, indent=2)
    
    if args.output:
        with open(args.output, 'w') as f:
            f.write(json_output)
        print(f"Data saved to {args.output}", file=sys.stderr)
    else:
        print(json_output)

if __name__ == "__main__":
    main()

File: scripts/fetch-matches.mjs
import axios from 'axios'
import dayjs from 'dayjs'
import fs from 'fs'
import path from 'path'

const COMPETITION_ID = 8
const SEASON_YEAR = 2025
const API_LIMIT = 100
const OUTPUT = `public/pl-form-comparison/${SEASON_YEAR}.json`
const OVERRIDE = process.env.OVERRIDE === 'true'
const UPCOMING_MATCHES_FILE = `scripts/resources/upcoming-matches.json`

async function fetchMatches(matchweek) {
  const url = `https://sdp-prem-prod.premier-league-prod.pulselive.com/api/v2/matches?competition=${COMPETITION_ID}&season=${SEASON_YEAR}&matchweek=${matchweek}&_limit=${API_LIMIT}`

  const headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
    Accept: 'application/json',
    'Accept-Language': 'en-US,en;q=0.9',
  }

  try {
    const response = await axios.get(url, { headers, timeout: 10000 })
    return response.data
  } catch (error) {
    console.error(`Error fetching matchweek ${matchweek}:`, error.message)
    return null
  }
}

function extractUpcomingMatches(jsonData) {
  const upcoming = {}

  for (const mw of jsonData.matchweeks) {
    for (const match of mw.data.data) {
      if (match.period === 'PreMatch') {
        const kickoff = new Date(match.kickoff.replace(' ', 'T') + 'Z')
        const kickoffUtc = kickoff.toISOString().split('T')[0] + 'T00:00:00Z'

        if (!upcoming[kickoffUtc]) {
          upcoming[kickoffUtc] = []
        }
        if (!upcoming[kickoffUtc].includes(mw.matchweek)) {
          upcoming[kickoffUtc].push(mw.matchweek)
        }
      }
    }
  }

  return upcoming
}

function classifyMatchweeksTime(upcoming) {
  const past = {}
  const future = {}
  for (const [date, weeks] of Object.entries(upcoming)) {
    if (dayjs(date).isAfter(dayjs())) {
      future[date] = weeks
    } else {
      past[date] = weeks
    }
  }
  return { past, future }
}

function ensureResourcesDir() {
  const dir = path.dirname(UPCOMING_MATCHES_FILE)
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true })
  }
}

async function main() {
  const upcomingMatchweeksRecord = JSON.parse(fs.readFileSync(UPCOMING_MATCHES_FILE, 'utf-8'))
  const { future, past } = classifyMatchweeksTime(upcomingMatchweeksRecord)
  const pastMatchweeks = new Set(Object.values(past).flat())

  console.info(`Fetching matchweeks: ${MATCHWEEKS.join(', ')}...`)
  const allData = []

  for (const week of pastMatchweeks) {
    console.info(`Fetching matchweek ${week}...`)
    const data = await fetchMatches(week)

    if (data !== null) {
      allData.push({
        matchweek: week,
        data: data,
      })
    } else {
      throw new Error(`Failed to fetch matchweek ${week}`)
    }
  }

  const result = {
    season: SEASON_YEAR,
    competition: COMPETITION_ID,
    matchweeks: allData,
  }

  let jsonOutput = result

  if (fs.existsSync(OUTPUT)) {
    const existingFile = fs.readFileSync(OUTPUT, 'utf-8')
    const existingJSON = JSON.parse(existingFile)

    const existingByWeek = Object.fromEntries(existingJSON.matchweeks.map((mw) => [mw.matchweek, mw]))

    result.matchweeks.forEach((mw) => {
      existingByWeek[mw.matchweek] = mw
    })

    jsonOutput.matchweeks = Object.values(existingByWeek).sort((a, b) => a.matchweek - b.matchweek)
  }

  fs.writeFileSync(OUTPUT, JSON.stringify(jsonOutput, null, 2))
  console.log(`Data saved to ${OUTPUT}`)

  ensureResourcesDir()

  fs.writeFileSync(UPCOMING_MATCHES_FILE, JSON.stringify(future, null, 2))
  console.log(`Upcoming matches saved to ${UPCOMING_MATCHES_FILE}`)
}

main().catch((error) => {
  console.error('Fatal error:', error)
  process.exit(1)
})
